src/librbd/io/SimpleSchedulerObjectDispatch.cc

```diff
@@ -0,0 +1,262 @@
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << object_len << dendl;
+
+  dispatch_delayed_requests(object_no);
+
+  return false;
+}
+
+template <typename I>
+bool SimpleSchedulerObjectDispatch<I>::compare_and_write(
+    const std::string &oid, uint64_t object_no, uint64_t object_off,
+    ceph::bufferlist&& cmp_data, ceph::bufferlist&& write_data,
+    const ::SnapContext &snapc, int op_flags,
+    const ZTracer::Trace &parent_trace, uint64_t* mismatch_offset,
+    int* object_dispatch_flags, uint64_t* journal_tid,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << cmp_data.length() << dendl;
+
+  dispatch_delayed_requests(object_no);
+
+  return false;
+}
+
+template <typename I>
+bool SimpleSchedulerObjectDispatch<I>::flush(
+    io::FlushSource flush_source, const ZTracer::Trace &parent_trace,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  dispatch_delayed_requests();
+
+  return false;
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests() {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  for (auto &it : m_delayed_requests) {
+    dispatch_delayed_requests(it.first, m_lock);
+  }
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    uint64_t object_no) {
+
+  Mutex::Locker locker(m_lock);
+  dispatch_delayed_requests(object_no, m_lock);
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+  uint64_t object_no, const Mutex &m_lock) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+
+  auto it = m_delayed_requests.find(object_no);
+  if (it == m_delayed_requests.end()) {
+    ldout(cct, 20) << "object_no=" << object_no << ": not found" << dendl;
+    return;
+  }
+
+  auto &delayed_requests = it->second;
+
+  ldout(cct, 20) << "object_no=" << object_no << ", "
+                 << delayed_requests.requests.size() << " requests"
+                 << dendl;
+
+  if (!delayed_requests.requests.empty()) {
+    auto requests = std::move(delayed_requests.requests);
+    auto ctx = new FunctionContext(
+      [this, requests](int r) {
+        for (auto on_dispatched : requests) {
+          on_dispatched->complete(r);
+        }
+      });
+
+    auto req = io::ObjectDispatchSpec::create_write(
+      m_image_ctx, io::OBJECT_DISPATCH_LAYER_SCHEDULER,
+      m_image_ctx->get_object_name(object_no), object_no,
+      delayed_requests.offset, std::move(delayed_requests.data),
+      delayed_requests.snapc, delayed_requests.op_flags, 0,
+      {} /* XXXMG */, ctx);
+    // XXXMG: req->object_dispatch_flags ?
```

Comment:
Yes -- you should pass the dispatch flags from the write request. Right now the flags can be merged. Perhaps add a helper ```can_merge_write``` method which can be expanded in the future. Plus, that method can assist w/ front/back merges as well.